### PR TITLE
[web] Send `flutterViewId` with platform view messages

### PR DESCRIPTION
**Pull request by mdebbar**
This PR prepares web platform views for the multi-view world.

Depends on https://github.com/flutter/engine/pull/46891
Part of https://github.com/flutter/flutter/issues/137287

### CODE REVIEW

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 1]
```diff
@@ -81,28 +85,34 @@ PlatformViewCreatedCallback? _createPlatformViewCallbackForElementCallback(
...
+  final int flutterViewId;
+
   bool _initialized = false;
 
   Future<void> _initialize() async {
     final Map<String, dynamic> args = <String, dynamic>{
-      'id': viewId,
-      'viewType': viewType,
+      'platformViewId': platformViewId,
+      'platformViewType': platformViewType,
```
Comment (ditman) [inline packages/flutter/lib/src/widgets/_html_element_view_web.dart:113]: This one can't land until the `engine` change lands. Especially changing the name of the parameter :)

File: packages/flutter/lib/src/widgets/view.dart [hunk 2]
```diff
@@ -131,7 +131,7 @@ class View extends StatelessWidget {
   ///
   /// See also:
   ///
-  ///  * [View.maybeOf], which throws instead of returning null if no
+  ///  * [View.maybeOf], which returns null instead of throwing if no
```
Comment (ditman) [inline packages/flutter/lib/src/widgets/view.dart:134]: The implementation of this method sounds like it'll throw if result is null (either by the code of the assert, or by returning `result!`? Revert this change?

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 3]
```diff
@@ -123,7 +133,11 @@ class _HtmlElementViewController extends PlatformViewController {
   @override
   Future<void> dispose() async {
     if (_initialized) {
-      await SystemChannels.platform_views.invokeMethod<void>('dispose', viewId);
+      final Map<String, dynamic> args = <String, dynamic>{
+        'platformViewId': viewId,
+        'viewId': flutterViewId,
+      };
+      await SystemChannels.platform_views.invokeMethod<void>('dispose', args);
```
Comment (ditman) [inline packages/flutter/lib/src/widgets/_html_element_view_web.dart:140]: I like this better, we should have some rule/lint against passing a scalar value as args for a system channel (at least we can give it a name with a `Map<String, Object?>`

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 4]
```diff
@@ -123,7 +133,11 @@ class _HtmlElementViewController extends PlatformViewController {
   @override
   Future<void> dispose() async {
     if (_initialized) {
-      await SystemChannels.platform_views.invokeMethod<void>('dispose', viewId);
+      final Map<String, dynamic> args = <String, dynamic>{
```
Comment (ditman) [inline packages/flutter/lib/src/widgets/_html_element_view_web.dart:136]:
```suggestion
      final Map<String, Object?> args = <String, Object?>{
```

**Review by ditman — COMMENTED** [review]
Minor comments, but in general LGTM!

(this can't land until the engine version does, though, since this is using new message argument names that are currently not understood by the engine)

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 5]
```diff
@@ -38,9 +39,10 @@ extension HtmlElementViewImpl on HtmlElementView {
   /// The implementation on Flutter Web builds an HTML platform view and handles
   /// its lifecycle.
   Widget buildImpl(BuildContext context) {
+    final FlutterView flutterView = View.of(context);
     return PlatformViewLink(
       viewType: viewType,
-      onCreatePlatformView: _createController,
+      onCreatePlatformView: (PlatformViewCreationParams params) => _createController(params, flutterView.viewId),
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:45]: I prefer to keep it as a lambda TBH, but I'll change it to a block syntax (instead of arrow function). The `surfaceFactory` parameter is also being passed as a lambda right below this.

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 6]
```diff
@@ -54,11 +56,13 @@ extension HtmlElementViewImpl on HtmlElementView {
   /// Creates the controller and kicks off its initialization.
   _HtmlElementViewController _createController(
     PlatformViewCreationParams params,
+    int viewId,
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:59]: `PlatformViewLink` and `PlatformViewCreationParams` are used by all platforms and I didn't want to step on their stuff. I also want to hear from @goderbauer about how other platforms are going to deal with this.

File: packages/flutter/lib/src/widgets/view.dart (hunk 2 above)
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/view.dart:134]: This comment is referring to `View.maybeOf` which doesn't throw.

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 3 above)
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:140]: Agreed!

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 7]
```diff
@@ -81,28 +85,34 @@ PlatformViewCreatedCallback? _createPlatformViewCallbackForElementCallback(
...
+    this.platformViewId,
+    this.platformViewType,
     this.creationParams,
+    this.flutterViewId,
   );
 
   @override
-  final int viewId;
+  int get viewId => platformViewId;
+  final int platformViewId;
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart]: Done

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:61]: > I didn't want to step on their stuff.

It's _our_ stuff. We own this all together. :)

Adding the viewId to PlatformViewCreationParams sounds reasonable to me. I imagine other platforms will likely also need to know what FlutterView a PlatformView is associated with. @cbracken @loic-sharma have you thought about this in the context of your Desktop Platform view work.

File: packages/flutter/lib/src/widgets/view.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/view.dart:134]: Thanks for catching this!

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:61]: One thing to note, though: A PlatformView could move from one view to another. When scrolling through this PR I didn't see anything that would update the view ID associated with the controller when this happens?

File: packages/flutter/test/widgets/view_utils.dart [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+extension WrapperlessWidgetTester on WidgetTester {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/view_utils.dart:8]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-extension

(I would just define this as a helper method until we're ready to actually implement this on WidgetTester itself)

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (ditman) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:61]: > A PlatformView could move from one view to another.

Hm, that's an interesting case. In the web at least, since we'd need to move the actual DOM Content for platform_view (because each view has its own DOM shadow root), it'd be equivalent to destroying the view and recreating it from the factory elsewhere (in most cases).

I'm not sure how the framework/engine handles this case.

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (ditman) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart]: I'm good with the lambda :P

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 1 above)
Comment (ditman) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart]: (The engine change has landed)

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (cbracken) [reply packages/flutter/lib/src/widgets/_html_element_view_web.dart:61]: I hadn't given this much thought I terms of desktop platform views but agreed we'll likely want this for desktop as well.

**Comment by ditman** [comment]
Wow, thanks for adding support for moving the flutter view!

The testing seems to be an analyzer issue:

`error â�¢ Missing concrete implementation of 'PlatformViewController.moveToFlutterView' â�¢ packages/flutter/test/services/fake_platform_views.dart:47:7 `

File: packages/flutter/lib/src/widgets/_html_element_view_web.dart [hunk 9]
```diff
@@ -120,10 +129,25 @@ class _HtmlElementViewController extends PlatformViewController {
...
   }
 
+  @override
+  Future<void> moveToFlutterView(int flutterViewId) async {
+    if (_initialized) {
+      final Map<String, Object?> args = <String, Object?>{
+        'platformViewId': platformViewId,
+        'viewId': flutterViewId,
+      };
+      await SystemChannels.platform_views.invokeMethod<void>('moveToFlutterView', args);
```
Comment (ditman) [inline packages/flutter/lib/src/widgets/_html_element_view_web.dart:139]: 👍 I'd call this `move` just to keep the same brevity of the `create`/`dispose` methods, but I don't have a strong opinion.

Comment (ditman) [inline packages/flutter/lib/src/widgets/_html_element_view_web.dart:139]:
```suggestion
      await SystemChannels.platform_views.invokeMethod<void>('move', args);
```

👍 I'd call this `move` just to keep the same brevity of the `create`/`dispose` methods, but I don't have a strong opinion.

**Comment by mdebbar** [comment]
I'm still working on tests.. my latest [commit](https://github.com/flutter/flutter/pull/136548/commits/21573b56c8a3b50eb5d1cb9224b79d98db72c40b) broke some tests.

Here's what I think so far: in tests, platform view creation happens synchronously in `PlatformViewLink.initState`, while in real code, the platform view creation happens asynchronously. With my recent change, platform view creation moved from `initState` to `build`, causing the tests to fail. @goderbauer any suggestions on how to deal with this? Or ideas on how to restructure the code to avoid this issue?

Note: The reason for moving platform view creation (or controller creation, really) from `initState` to `build` is that we need `View.of(context)` now. Adding inherited widget dependencies isn't allowed in `initState` apparently (I'm hitting [this error](https://github.com/flutter/flutter/blob/5dd2a4e0aaef0072786316e2f01b1f8631c7d388/packages/flutter/lib/src/widgets/framework.dart#L5705-L5705)).

**Comment by ditman** [comment]
> The reason for moving platform view creation (or controller creation, really) from `initState` to `build` is that we need `View.of(context)` now.

Ahhh, tricky tricky :/

File: packages/flutter/lib/src/services/platform_views.dart [hunk 10]
```diff
@@ -1480,6 +1480,9 @@ abstract class PlatformViewController {
   /// an initial position.
   Future<void> create({Size? size, Offset? position}) async {}
 
+  /// Moves the platform view to a new [FlutterView].
+  Future<void> moveToFlutterView(int flutterViewId) async {}
```
Comment (loic-sharma) [inline packages/flutter/lib/src/services/platform_views.dart:1484]: Could we update the `PlatformViewController.viewId` documentation on lines 1451-1457 to clarify that identifier is for the platform view (instead of a Flutter view)?

**Comment by goderbauer** [comment]
Missed this ping last week. Apologies.

Instead of moving it from `initState` to `build` could you move it to `didChangeDependencies`? That one runs right after initState and again whenever the dependencies change.

**Comment by goderbauer** [comment]
(triage): @mdebbar said this is still on his radar in the next few weeks.

**Comment by yjbanov** [comment]
Attempting to look at it from 30000 feet: Can `viewId` be determined by the slot created using `SceneBuilder.addPlatformView`? Then we don't need extra API on the platform view creation side. A platform view can start as "detached", and it would be "attached" when the framework calls `FlutterView.render` passing it a `Scene` containing the platform view ID. To move the platform view to a different `FlutterView`, simply call `render` on the new `FlutterView` and pass the platform view ID there. The engine can `assert` that the same platform view does not appear in multiple `FlutterView`s simultaneously.

**Comment by mdebbar** [comment]
Update: this PR may become superseded by https://github.com/flutter/engine/pull/48960
